### ISSUE TITLE
lint: no single pass gives both human-readable output and JSON, and --json carries no finding text

### ISSUE THREAD

**Issue by vlad-ryzhkov** [zernie/agent-lint#182]
**Version:** vigiles 21.0.1

## Summary

There is no way to get the human-readable output and the JSON report from one `lint` run, and `--json` carries no finding text. A CI job that wants both has to scan the repo twice.

## What we hit

Our PR gate needs two things from the same scan:

- the human-readable output, because that is what a reader opens from the job log — it names the skills and explains each finding;
- the JSON, because we build a sticky PR comment from the counters.

`--json` replaces the human-readable output rather than adding to it, and the flag list has no third option:

```
$ npx vigiles lint --help
  vigiles lint  [files...]   Gate it in CI. The same checks — but a finding fails the build.

Flags: --catalog-only --harness= --help --json --max-rules= --summary
```

`--summary` is not a substitute — on our repo it prints one line:

```
vigiles: 35 untested surfaces
```

So the job runs the full scan twice. On our repo that is 57 files and about 3.2 s per pass — small, but it is the same work done twice, and it will not stay small as the corpus grows.

## The second half: `--json` has no finding text

The JSON report is 58 integer counters plus a `files` array of paths. There are no messages, no per-finding paths, no rule ids attached to a location. Checked on our repo:

| probe | present in `--json` |
| --- | --- |
| `"disallowed-tools"` | no |
| `"budget 500"` | no |
| `"no test or eval"` | no |
| a skill name from a finding | only inside `files[]`, not attached to any finding |

So the JSON cannot reconstruct what the human-readable run said. That is why we cannot drop the second pass and render the log from the JSON ourselves.

## What would help

Either one of these solves it:

1. **`--json-out=<file>`** — write the JSON to a file while stdout keeps the human-readable output. One scan, both artefacts, no change to existing behaviour.
2. **Findings inside `--json`** — an array of `{ rule, severity, path, message }` beside the counters. This is the more useful of the two for us: it would let a consumer build its own report and stop hand-maintaining label text per counter key. Today we keep a hardcoded map from `descriptionBudgetIssues` to "descriptions over the 500-char budget", which silently goes stale when a rule is renamed.

Both together would be ideal, but `--json-out` alone would already remove the duplicate scan.

**Comment by zernie** [comment]
Half fixed and merged in #187 (`85c54e7e`). Saying which half up front, because you ranked them and I did the one you ranked second.

## Shipped: `--json-out=` (your option 1)

```bash
npx vigiles lint . --json-out=reports/lint.json
```

stdout stays the human-readable run, the JSON lands on disk, one scan. That removes the duplicate pass you measured at ~3.2 s and growing.

**Also relevant to your first half:** the JSON now carries `totals: { findings, errors, warnings }` — the same numbers the log's new summary line prints, from one computed source (#183). It does not give you finding text, but it does mean the counters you build the sticky comment from no longer need reconciling against the log.

## NOT shipped: findings inside `--json` (your option 2)

You called it the more useful of the two and you are right — a `{ rule, severity, path, message }` array is what lets a consumer stop hand-maintaining a label map that goes stale on a rename. It is not in this change.

The honest reason: every check currently *prints* its findings and returns a count. Twenty-odd of them, each with its own message construction inline at the console call. Turning that into returned data is a real refactor of the whole check layer, not an addition to it — and doing it badly would mean two sources of message text, which is the drift you are already describing at the label-map level.

So it is a separate piece of work rather than something to squeeze into a PR that changes five other things. **Worth its own issue if you want it tracked** — I would rather you filed it with your use case attached than have me file a thinner version on your behalf.

What you can build on today, if it helps in the meantime: `totals` for the headline number, the per-rule counters for the breakdown, and `--json-out` so neither costs a second scan.

---
_Generated by [Claude Code](https://claude.ai/code)_
